### DISCUSSION TITLE
Creating an ODataClient at runtime in browser

### DISCUSSION THREAD

**Discussion by tobikugel** [odata2ts/odata2ts#313] [Q&A]
Hello,
I'm working on a browser-based coding sandbox project where users can choose an OData service at runtime, and I want to provide them an OData client (generated by your library) to have a type-safe way of accessing the OData service.
At this point, thank you for providing your cool library!

Now more about my problem:
The way I understand it, I need to first load the OData service's metadata file, then run the npm run gen-odata command to generate the appropriate data model and client service. However, since I would like to do that in my browser, executing binaries from within the browser isn't possible, I'm looking for an alternative solution to generate these models directly in the browser.

Has anyone faced a similar situation, or can anyone suggest alternative approaches to generating an OData client entirely in the browser? I'm hoping to avoid moving this process to a separate backend service.

**Comment by texttechne** [comment]
Hey,

### Static OData Services
first off, are these OData services you are targeting moving targets? In other words:
* Do you want to target a fixed list of previously known OData services?
* If so, are these OData services fully developed?

If you can check both boxes you don't need the generator at runtime.
You just use the generated stuff in your webapp, well, like usual...

### Dynamic Generation
Regarding the generator itself, it's intended to work in a node.js environment.
Your challenges here would be:
* simulate the file system
* all dependencies of the generator package must work in your browser-based environment
* the generator is pure ESM which will force you to also use ESM in your app in the end

I would suggest to research one of the following approaches:
* Make it an Electron app: browser-based, but must be installed (like Spotify or MS Teams)
* Maybe service workers are a means to solve this 
  * PWA might also be of interest, but at least demo quite nicely what service workers can do

**Reply by tobikugel** [reply]
My needs are exactly the opposite of the first two points.
Hence, the list of needed services is unknown and the OData services (EDMX-files) are likely subject to changes and updates.

Thank you for your thoughtful answer and giving me more insights. 
Making the app based on an Electron app or PWA is also not an option.

I will continue digging deeper into this topic and eventually share my approach.

**Reply by tobikugel** [reply]
I haven't found a solution to the requirements. 
Therefore what I did was setting up a node.js server, creating the files and bundling them on it, and serving the bundles to the frontend upon request.

**Reply by texttechne** [reply]
What's the argument against [service workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API)? Why wouldn't that approach work?
